pkg/adapters/rules/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,284 @@
+var _ a8api.RulesService = (*Adapter)(nil)
+
+// Package global logger
+var logger = logging.GetLogger("KUBERNETES_RULES")
+
+// Config stores configurable attributes of the Kubernetes adapter.
+type Config struct {
+	URL       string
+	Token     string
+	Namespace auth.Namespace
+}
+
+// Adapter for Kubernetes Service Discovery.
+type Adapter struct {
+
+	// kubernetes REST client
+	client *rest.RESTClient
+
+	// workqueue is used to queue and process events send from the cache controllers
+	workqueue *kubepkg.Workqueue
+
+	// rulesCache caches rules from Kubernetes API
+	rulesCache      cache.Store
+	rulesController cache.ControllerInterface
+
+	// rules maps a rule unique name to a api rule object.
+	rules map[string]*a8api.Rule
+
+	// namespace from which to sync rules
+	namespace string
+
+	// stopChan for stop signals
+	stopChan chan struct{}
+
+	// revision . We assume that sidecar and rules fetcher runs always together
+	revision int
+
+	//ruleChan chan struct{}
+	// mutex is used to synchronize access to the 'services' map,
+	// which is read by the ListXXX() methods (externally),
+	// and is written by the cache event handlers (internally).
+	// Given that we expect a single reader only, we use a regular sync.Mutex rather than a sync.RWMutex.
+	mutex sync.Mutex
+}
+
+// New creates and starts a new Kubernetes Rules adapter.
+func New(config Config) (*Adapter, error) {
+	logger.Info("Adapter starting..." )
+	namespace := config.Namespace.String()
+	// If no namespace is specified, fallback to default namespace
+	if namespace == "" {
+		namespace = "default"
+	}
+
+	// Create the kubernetes client for the rules resource
+	client, err := createRulesClient(config)
+	if err != nil {
+		return nil, err
+	}
+	workqueue := kubepkg.NewWorkqueue()
+
+	adapter := &Adapter{
+		client:    client,
+		workqueue: workqueue,
+		rules:     make(map[string]*a8api.Rule),
+		namespace: namespace,
+		revision: 0,
+	}
+	adapter.rulesCache, adapter.rulesController = cache.NewInformer(
+		cache.NewListWatchFromClient(client, fmt.Sprintf("%ss", strings.Replace(ResourceName, "-", "", -1)), namespace, nil),
+		&RoutingRule{},
+		RulesCacheResyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    workqueue.EnqueueingAddFunc(adapter.addRule),
+			UpdateFunc: workqueue.EnqueueingUpdateFunc(adapter.updateRules),
+			DeleteFunc: workqueue.EnqueueingDeleteFunc(adapter.deleteRules),
+		},
+	)
+	logger.Info("Calling Adapter start" )
+	adapter.workqueue.Start()
+
+	return adapter, adapter.Start()
+}
+
+// Start synchronizing the Kubernetes adapter.
+func (a *Adapter) Start() error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	if a.stopChan != nil {
+		err := fmt.Errorf("kubernetes adapter already started")
+
+		logger.WithError(err).Errorf("Failed starting Kubernetes adapter")
+		return err
+	}
+	a.stopChan = make(chan struct{})
+
+	go a.rulesController.Run(a.stopChan)
+	return nil
+}
+
+func createRulesClient(config Config) (*rest.RESTClient, error) {
```

Comment:
This should probably re-use the existing `kubepkg.NewTPRClient()` function.
See [here](https://github.com/amalgam8/amalgam8/blob/master/cmd/k8srules/controller.go#L75) for example (but need to pass-in the url/token if present).